lib/Adapter/Tolerant/TolerantIndexBuilder.php

```diff
@@ -0,0 +1,155 @@
+<?php
+
+namespace Phpactor\Indexer\Adapter\Tolerant;
+
+use Microsoft\PhpParser\Node;
+use Microsoft\PhpParser\Node\QualifiedName;
+use Microsoft\PhpParser\Node\Statement\ClassDeclaration;
+use Microsoft\PhpParser\Node\Statement\FunctionDeclaration;
+use Microsoft\PhpParser\Parser;
+use Phpactor\Indexer\Model\Index;
+use Phpactor\Indexer\Model\IndexBuilder;
+use Phpactor\Indexer\Model\Record\ClassRecord;
+use Phpactor\Indexer\Model\Record\FunctionRecord;
+use Phpactor\Name\FullyQualifiedName;
+use Phpactor\TextDocument\ByteOffset;
+use SplFileInfo;
+
+class TolerantIndexBuilder implements IndexBuilder
+{
+    /**
+     * @var Index
+     */
+    private $index;
+
+    /**
+     * @var Parser
+     */
+    private $parser;
+
+    public function __construct(
+        Index $index,
+        ?Parser $parser = null
+    ) {
+        $this->index = $index;
+        $this->parser = $parser ?: new Parser();
+    }
+
+    public function index(SplFileInfo $info): void
+    {
+        $contents = @file_get_contents($info->getPathname());
+
+        if (false === $contents) {
+            return;
+        }
+
+        $node = $this->parser->parseSourceFile($contents, $info->getPathname());
+        $this->indexNode($info, $node);
+    }
+
+    public function done(): void
+    {
+        $this->index->updateTimestamp();
+    }
+
+    private function indexNode(SplFileInfo $info, Node $node): void
+    {
+        if ($node instanceof ClassDeclaration) {
+            $this->indexClassDeclaration($info, $node);
+            return;
+        }
+
+        if ($node instanceof FunctionDeclaration) {
+            $this->indexFunction($info, $node);
+            return;
+        }
+
+        foreach ($node->getChildNodes() as $childNode) {
+            $this->indexNode($info, $childNode);
+        }
+    }
+
+    private function indexClassDeclaration(SplFileInfo $info, ClassDeclaration $node): void
+    {
+        $record = $this->index->get(ClassRecord::fromName($node->getNamespacedName()->getFullyQualifiedNameText()));
+        assert($record instanceof ClassRecord);
+        $record->withLastModified($info->getCTime());
+        $record->withStart(ByteOffset::fromInt($node->getStart()));
+        $record->withType('class');
+        $record->withFilePath($info->getPathname());
+
+        // de-reference this class
```

Comment:
```suggestion
        // de-reference this class in case of changes since last indexation
```

I was wondering why we did this until I read `indexClassInterfaces` and start wondering why on earth we would remove the reference to recreate them just after :)

Maybe even extract it in a method to make gave a clear name while hiding the "confusing" implementation ?
`clearPreviouslyIndexedImplementations` or something like that, I'm not great with names so they always end up 3 foot long :D

Reply:
Extracted method and added comment